### PR TITLE
allow for retry on typically transient k8s errors in both core controller and resolver for remote resolution

### PR DESCRIPTION
**Pull request by gabemontero**
# Changes

collaborating with @khrm here is an augmented version of https://github.com/tektoncd/pipeline/pull/7893

Fixes https://github.com/tektoncd/pipeline/issues/7909

During both sides of remote resolution (core controller and resolver) typically transient kubernetes errors were being treated as permanent knative errors and no attempts at trying to reconcile again were made, leading to failures which could be avoided.

These changes addresses that.

An example log snippet from the core controller

```
Pipeline rh-acs-tenant/operator-on-pull-request-bwqxj can't be Run; it contains Tasks that don't exist: Couldn't retrieve Task "": retryable error validating referenced object source-build: Internal error occurred: failed calling webhook "validation.webhook.pipeline.tekton.dev": failed to call webhook: Post "https://tekton-pipelines-webhook.openshift-pipelines.svc:443/resource-validation?timeout=10s": context deadline exceeded
```

Accompanying log snippet from the resolver

```
{"level":"error","ts":"2024-04-17T10:50:05.866Z","logger":"controller","caller":"controller/controller.go:566","msg":"Reconcile error","commit":"f0a1d64","knative.dev/traceid":"b893d6a6-2eb7-4a53-b502-1348803a7085","knative.dev/key":"rh-acs-tenant/bundles-780a1fe396cb0f8c702b34e9289fc770","duration":"10.3628985s","error":"error updating resource request \"rh-acs-tenant/bundles-780a1fe396cb0f8c702b34e9289fc770\" with data: Internal error occurred: failed calling webhook \"webhook.pipeline.tekton.dev\": failed to call webhook: Post \"https://tekton-pipelines-webhook.openshift-pipelines.svc:443/defaulting?timeout=10s\": context deadline exceeded","stacktrace":"knative.dev/pkg/controller.(*Impl).handleErr\n\t/go/src/github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:566\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\t/go/src/github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:543\nknative.dev/pkg/controller.(*Impl).RunContext.func3\n\t/go/src/github.com/tektoncd/pipeline/vendor/knative.dev/pkg/controller/controller.go:491"}
```

/kind bug

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ n/a] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [/ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ /] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [/ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [/] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ /] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [n/a ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
This fix address the lack of retry on transient kubernetes errors during remote resolution for tasks, etc. 
```

### CODE REVIEW

**Comment by gabemontero** [comment]
just pushed an extra-credit commit that properly generates `TaskNotFoundError` when the task ref name is not set in the `Name` field, but rather comes from the `name` Param

like:

```
      taskRef:
        kind: Task
        params:
        - name: name
          value: summary
        - name: bundle
          value: myregistry/myrepo/myimage:mytag_or_sha
        - name: kind
          value: task
        resolver: bundles
```

I can break it off into a separate PR if desired

**Comment by savitaashture** [comment]
LGTM

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 1]
```diff
@@ -645,8 +645,18 @@ func resolveTask(
 			case errors.Is(err, remote.ErrRequestInProgress):
 				return rt, err
 			case err != nil:
+				name := pipelineTask.TaskRef.Name
```
Comment (gabemontero) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:648]: so if you look a few lines up @khrm @savitaashture at line 643:

`t, _, vr, err := getTask(ctx, pipelineTask.TaskRef.Name)`

I am unclear on how this would ever work if `pipelineTask.TaskRef.Name` is an empty string.  Maybe the name from the param is seeded in the `getTask` func when it is constructed and the `pipelineTask.TaskRef.Name` is ignored?  I did not have time to trace through the code and find the image bundle version of `getTask` to sort that out.

I'm going to try to nail that down today.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 2]
```diff
@@ -645,8 +645,18 @@ func resolveTask(
...
+				if len(strings.TrimSpace(name)) == 0 {
+					for _, p := range pipelineTask.TaskRef.Params {
+						// TODO: amazingly, there appears to be no constants for the resolver related param names
```
Comment (gabemontero) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:651]: I finally found the constants for these over at https://github.com/tektoncd/pipeline/blob/main/pkg/resolution/resolver/bundle/params.go

Also found a utility function that I could reuse to get the name from the params

/hold

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 1 above)
Comment (gabemontero) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:648]: OK the amount of layering can make one dizzy, but *eventually* we land at https://github.com/tektoncd/pipeline/blob/main/pkg/resolution/resolver/bundle/params.go#L79 and https://github.com/tektoncd/pipeline/blob/main/pkg/resolution/resolver/bundle/bundle.go#L112

it uses the params and ignore the name for image bundles

I'm going to try and see about using  `OptionsForParams` over in https://github.com/tektoncd/pipeline/blob/main/pkg/resolution/resolver/bundle/params.go#L43

Comment (gabemontero) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:648]: wait .... parms can vary by resolution type .... need to research more

**Review by chitrangpatel — APPROVED** [review]
Thanks!

File: pkg/resolution/resource/name.go [hunk 3]
```diff
@@ -69,3 +79,25 @@ func GenerateDeterministicName(prefix, base string, params v1.Params) (string, e
...
+				name = string(asJSON)
+			}
+			paramString += fmt.Sprintf("name = %s\n", name)
+		}
+		if p.Name == ParamURL {
+			paramString += fmt.Sprintf("url = %s\n", p.Value.StringVal)
+		}
+	}
+	return paramString
+}
```
Comment (chitrangpatel) [inline pkg/resolution/resource/name.go:103]: Can we add tests for this exported function? And also some doc string? Thanks!

Comment (gabemontero) [reply pkg/resolution/resource/name.go:106]: updated - thanks !

Comment (chitrangpatel) [reply pkg/resolution/resource/name.go:106]: Thanks!

**Comment by gabemontero** [comment]
ok @chitrangpatel @afrittoli I have rebased on top of @chitrangpatel 's recently merged refactor

**Review by chitrangpatel — COMMENTED** [review]
Can you also make the corresponding changes from `pkg/resolution/resolver/framework/reconciler.go` and its test

into `pkg/remoteresolution/resolver/framework/reconciler.go` and its test file?

Thats the newer framework. The current one will be deprecated at somepoint.

**Comment by afrittoli** [comment]
> Can you also make the corresponding changes from `pkg/resolution/resolver/framework/reconciler.go` and its test
> 
> into `pkg/remoteresolution/resolver/framework/reconciler.go` and its test file?
> 
> All the other files are LGTM!
> 
> Thats the newer framework. The current one will be deprecated at somepoint.

Thanks @chitrangpatel - as I mentioned on slack, I think we should *only* do the changes in the new framework, since that's what is actually deployed and used now, and I would not want to have to maintain both. The old framework is still around to provide some room for potential users to move to the new one, but I would not backport changes to it.

@chitrangpatel should we mark the old resolvers as deprecated already?

**Comment by chitrangpatel** [comment]
> > Can you also make the corresponding changes from `pkg/resolution/resolver/framework/reconciler.go` and its test
> > into `pkg/remoteresolution/resolver/framework/reconciler.go` and its test file?
> > All the other files are LGTM!
> > Thats the newer framework. The current one will be deprecated at somepoint.
> 
> Thanks @chitrangpatel - as I mentioned on slack, I think we should _only_ do the changes in the new framework, since that's what is actually deployed and used now, and I would not want to have to maintain both. The old framework is still around to provide some room for potential users to move to the new one, but I would not backport changes to it.
> 
> @chitrangpatel should we mark the old resolvers as deprecated already?

Yes, I'm happy to. I can submit a PR to do that today.

**Comment by gabemontero** [comment]
ok @chitrangpatel @afrittoli I moved those changes from the deprecated resolution subpackage over to remoteresolution

there was a minor goimport cleanup I left in the old deprecated package reconciler_test.go file

I also squashed the commits and updated the git commit message accordingling

PTAL / thanks

**Review by vdemeester — APPROVED** [review]

File: pkg/resolution/resource/name.go [hunk 4]
```diff
@@ -111,3 +121,28 @@ func GenerateDeterministicNameFromSpec(prefix, base string, resolutionSpec *v1be
...
+	for _, p := range params {
+		if p.Name == ParamName {
+			name := p.Value.StringVal
+			if p.Value.Type != v1.ParamTypeString {
+				asJSON, err := p.Value.MarshalJSON()
+				if err != nil {
+					paramString += fmt.Sprintf("name could not be marshalled: %s\n", err.Error())
+					continue
+				}
+				name = string(asJSON)
```
Comment (afrittoli) [inline pkg/resolution/resource/name.go:139]: NIT: coverage shows no coverage for this line, can we add a test case with an object param that can be marshalled as JSON? It can be a separate PR.

**Review by afrittoli — COMMENTED** [review]
Thanks for this PR!
I added some minor comments, but nothing blocking.
/lgtm

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go [hunk 5]
```diff
@@ -645,8 +646,12 @@ func resolveTask(
 			case errors.Is(err, remote.ErrRequestInProgress):
 				return rt, err
 			case err != nil:
+				name := pipelineTask.TaskRef.Name
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:649]: part of  next push

[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:649]: oops .... I pushed the rebase before this update @afrittoli and with @chitrangpatel lgtm the pr has merged

I'll handle this in and the unit test coverage item in a follow up PR I should be able to open today.

File: pkg/resolution/resource/name.go (hunk 4 above)
Comment (gabemontero) [reply pkg/resolution/resource/name.go:144]: Yep I'll handle this in separate PR ... hopefully should have it open today

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:649]: see https://github.com/tektoncd/pipeline/pull/7950

File: pkg/resolution/resource/name.go (hunk 4 above)
Comment (gabemontero) [reply pkg/resolution/resource/name.go:144]: see https://github.com/tektoncd/pipeline/pull/7950
